src/tiledb/cloud/file/indexing.py

```diff
@@ -0,0 +1,405 @@
+from enum import Enum
+from typing import Any, Dict, List, Mapping, Optional, Sequence
+
+import tiledb
+from tiledb.cloud import dag
+from tiledb.cloud.utilities import as_batch
+from tiledb.cloud.utilities import get_logger_wrapper
+
+
+class IndexTypes(str, Enum):
+    FLAT = "FLAT"
+    IVF_FLAT = "IVF_FLAT"
+    VAMANA = "VAMANA"
+
+
+# ===============================
+#             UDFs
+# ===============================
+def create_dataset_udf(
+    search_uri: str,
+    index_uri: str,
+    *,
+    config: Optional[dict] = None,
+    environment_variables: Optional[Mapping[str, str]] = None,
+    verbose: bool = False,
+    # Index creation params
+    index_type: IndexTypes = IndexTypes.IVF_FLAT,
+    index_creation_kwargs: Optional[Dict] = None,
+    # DirectoryTextReader params
+    pattern: str = "*",
+    ignore: Optional[Sequence[str]] = ("[.]*", "*/[.]*"),
+    suffixes: Optional[Sequence[str]] = None,
+    max_files: Optional[int] = None,
+    text_splitter: str = "RecursiveCharacterTextSplitter",
+    text_splitter_kwargs: Optional[Dict] = None,
+    # Embedding params
+    embedding_class: str = "LangChainEmbedding",
+    embedding_kwargs: Optional[Dict] = None,
+) -> str:
+    """
+    Create a TileDB vector search dataset.
+    """
+    import tiledb.vector_search as vs
+    import tiledb.vector_search.embeddings as embeddings_module
+    from tiledb.vector_search.object_api import object_index
+    from tiledb.vector_search.object_readers import DirectoryTextReader
+
+    logger = get_logger_wrapper(verbose)
+    logger.debug("tiledb-vector-search=%s", vs.__version__)
+
+    # Check if the dataset already exists
+    with tiledb.scope_ctx(config):
+        if environment_variables is None:
+            environment_variables = {}
+        if index_creation_kwargs is None:
+            index_creation_kwargs = {}
```

Comment:
`environment_variables = environment_variables or {}` here. (this construction won’t work for the others below but will make these shorter)